tests/src/metadata.rs

```diff
@@ -0,0 +1,188 @@
+        validate_hints = get_flag_metadata(line, "Hints").or(validate_hints);
+        validate_autocomplete =
+            get_flag_metadata(line, "Autocomplete").or(validate_autocomplete);
+
+        fn num(s: &mut Scanner) -> Option<isize> {
+            let mut first = true;
+            let n = &s.eat_while(|c: char| {
+                let valid = first && c == '-' || c.is_numeric();
+                first = false;
+                valid
+            });
+            n.parse().ok()
+        }
+
+        let comments_until_code =
+            lines[i..].iter().take_while(|line| line.starts_with("//")).count();
+
+        let pos = |s: &mut Scanner| -> Option<usize> {
+            let first = num(s)? - 1;
+            let (delta, column) =
+                if s.eat_if(':') { (first, num(s)? - 1) } else { (0, first) };
+            let line = (i + comments_until_code).checked_add_signed(delta)?;
+            source.line_column_to_byte(line, usize::try_from(column).ok()?)
+        };
+
+        let range = |s: &mut Scanner| -> Option<Range<usize>> {
+            if s.eat_if("-1") {
+                let mut add = 1;
+                while let Some(line) = lines.get(i + add) {
+                    if !line.starts_with("//") {
+                        break;
+                    }
+                    add += 1;
+                }
+                let next_line = lines.get(i + add)?;
+                let col = next_line.chars().count();
+
+                let index = source.line_column_to_byte(i + add, col)?;
+                return Some(index..index);
+            }
+            let start = pos(s)?;
+            let end = if s.eat_if('-') { pos(s)? } else { start };
+            Some(start..end)
+        };
+
+        for kind in AnnotationKind::iter() {
+            let Some(expectation) = get_metadata(line, kind.as_str()) else { continue };
+            let mut s = Scanner::new(expectation);
+            let range = range(&mut s);
+            let rest = if range.is_some() { s.after() } else { s.string() };
+            let message = rest
+                .trim()
+                .replace("VERSION", &PackageVersion::compiler().to_string())
+                .into();
+            annotations.insert(Annotation { kind, range, message });
+        }
+    }
+
+    TestPartMetadata {
+        part_configuration: TestConfiguration {
+            compare_ref,
+            validate_hints,
+            validate_autocomplete,
+        },
+        annotations,
+    }
+}
+
+pub fn get_metadata<'a>(line: &'a str, key: &str) -> Option<&'a str> {
+    line.strip_prefix(eco_format!("// {key}: ").as_str())
+}
+
+pub fn get_flag_metadata(line: &str, key: &str) -> Option<bool> {
+    get_metadata(line, key)
+        .map(|value| {
+            if !(value == "true" || value == "false") {
+            println!("WARNING: invalid use of get_flag_metadata: flag should be `true` or `false` but is `{value}`");
```

Comment:
Could write to output instead, would modify the function signature but could be worth it, I don't mind doing it